bin/rm/rm.c

```diff
@@ -338,7 +338,7 @@ rm_file(char **argv)
 			if (Wflag) {
 				sb.st_mode = S_IFWHT|S_IWUSR|S_IRUSR;
 			} else {
-				if (!fflag || errno != ENOENT) {
+				if (!fflag && errno != ENOENT) {
```

Comment:
@stesser basically, either:
```c
if (!fflag || (errno != ENOENT && errno != EINVAL)) {
```
Or better, modify the msdos file system to accept `*` and `?`. The latter sounds like a better approach, although I have no idea how to do it. I'll have to look it up with it over the weekend. Thank you.

Reply:
I can modify the msdos file system to return ENOENT for invalid file names, since I know that code after fixing a few long standing bugs:
```
diff --git a/sys/fs/msdosfs/msdosfs_lookup.c b/sys/fs/msdosfs/msdosfs_lookup.c
index 2a90339d0878..9478196846d2 100644
--- a/sys/fs/msdosfs/msdosfs_lookup.c
+++ b/sys/fs/msdosfs/msdosfs_lookup.c
@@ -88,9 +88,13 @@ msdosfs_lookup_checker(struct msdosfsmount *pmp, struct vnode *dvp,
 int
 msdosfs_lookup(struct vop_cachedlookup_args *ap)
 {
+       int error;
 
-       return (msdosfs_lookup_ino(ap->a_dvp, ap->a_vpp, ap->a_cnp, NULL,
-           NULL));
+       error = msdosfs_lookup_ino(ap->a_dvp, ap->a_vpp, ap->a_cnp, NULL,
+           NULL);
+       if (error == EINVAL)
+               error = ENOENT;
+       return (error);
 }
 
 struct deget_dotdot {
```
The result is as expected, `rm -f` succeeds without reporting an error:
```
# rm -f /boot/efi/foo*
# echo $?
0
```
But this patch does also affect, e.g., `ls -l`:
```
# ls -l /boot/efi/foo*
ls: /boot/efi/foo*: No such file or directory
```
Without the patch, the result has been `ls: /boot/efi/foo*: Invalid argument`, and the behavior with the patch is similar to what would be returned on UFS.

But another side effect is that `mv` also returns `No such file or directory`instead of `Invalid argument` when the target file name is invalid:
```
# touch /boot/efi/A.DAT
# mv /boot/efi/A.DAT /boot/efi/B*.DAT
mv: rename /boot/efi/A.DAT to /boot/efi/B*.DAT: No such file or directory
```
This is unexpected and might break existing code that uses `rename()` or scripts that check the result of `mv`!
On UFS, renaming to `B*.DAT`would work and would result in a file with an asterisk in its name (as it is valid in filenames on UFS).

Reply:
Thank you! At a first glance, I like it (even for `mv`)!